e2e_context_test.go

```diff
@@ -393,3 +393,45 @@ func TestContextPerRequestWithTimeoutCancelledByContext(t *testing.T) {
 	// expected error should occur
 	assert.True(t, suppressor.expErrorOccurred)
 }
+
+func TestContextPerRequestRetryCancelledByContext(t *testing.T) {
+	var callCount int
+	var isConfigCtxCancelled bool
+	delayDuration := TimeOutDuration / 2
+
+	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		callCount++
+		w.WriteHeader(http.StatusInternalServerError)
+	}))
+	defer ts.Close()
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	go func(fn context.CancelFunc) {
+		time.Sleep(delayDuration / 2)
+		fn()
+		isConfigCtxCancelled = true
+	}(cancel)
+
+	// Config with context cancelled error
+	suppressor := newExpErrorSuppressor(t,
+		func(err error) bool {
+			return strings.Contains(err.Error(), context.Canceled.Error())
+		})
+
+	e := WithConfig(Config{
+		BaseURL:          ts.URL,
+		AssertionHandler: suppressor,
+	})
+
+	e.GET("/").
+		WithContext(ctx).
+		WithMaxRetries(1).
+		WithRetryPolicy(RetryAllErrors).
+		WithRetryDelay(delayDuration, TimeOutDuration).
```

Comment:
So here is the timing:

* Request.Expect sends request, receives error
* Request.Expect sleeps for 500ms
* during this sleep, after 250ms, we cancel context
* we assume that Request.Expect will see the cancelled context before timeout expires (i.e. in less then 250ms)

This makes sense and is usually OK, but sometimes on slow CI workers there are big delays. So usually it's better to avoid tests that depend on short timeouts.

I think here we can just use some very large timeout like a minute or 10 minutes, because the test anyway does not wait for the full timeout, right? This way we can be sure that the test will work even when CI is very slow.

If we set very large timeout, e.g. 10 minutes, we can also add a check that the actual wait time was significantly smaller (e.g. less then 5 minutes).

Reply:
> This makes sense and is usually OK, but sometimes on slow CI workers there are big delays. So usually it's better to avoid tests that depend on short timeouts.

hi, sry but could you elaborate more about this big delays? where will this big delays happen? 

and what will it cause? (could it _somehow_ makes the 500ms comes earlier than 200ms)

because I think at the end 250ms will come earlier than 500ms right

many thanks!

Reply:
In general, if you run two goroutines:

```go
go func() {
    time.Sleep(time.Second)
    println("g1")
}()

go func() {
    time.Sleep(time.Second * 2)
    println("g2")
}()
```

The following is guaranteed:

* g1 will be printed *at least* after 1 second
* g2 will be printed *at least* after 2 seconds

Nothing more. Here are some examples of what is NOT guaranteed:

* g1 will be printed exactly after 1 second
* g1 will be printed before g2

Normally, the two assertions above are true. However, on a loaded system, when there are more ready goroutines than available CPUs, both prints may be delayed and even reordered, depending on how overloaded are goroutine queues and how they will be scheduled.

So it's better to avoid relying on real time when possible and don't assume that everything will be scheduled in time and in order.

Reply:
got it

will increase the time

thanks for the explanation